### PR TITLE
Add VirtualServer nodejs example client

### PR DESCRIPTION
**Pull request by YitzyD**
closes: https://github.com/coreweave/virtual-server-operator/issues/8

### CODE REVIEW

**Comment by YitzyD** [comment]
Should I add a readme?

File: virtual-server/examples/nodejs/virtual-server-client-examples/main.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
+const VSClient = require("./client.js")
+
+// A sample VirtualServer manifest
+const virtualServerManifest = `
```
Comment (salanki) [inline virtual-server/examples/nodejs/virtual-server-client-examples/main.js:4]: This should be a JSON object, submitting a YAML string is not very programattic.

**Review by salanki — CHANGES_REQUESTED** [review]

File: virtual-server/examples/nodejs/virtual-server-client-examples/main.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    .then(o => {
+      if(o.statusCode === 202) {
+        started = true
+        console.log("VS started")
+      }
+    })
+    .catch(err => err.statusCode === 404 ? console.log("Waiting for subresource API") : console.log(err.toString())) 
+  }
+  
+  // Get the VirtualServer we created
```
Comment (salanki) [inline virtual-server/examples/nodejs/virtual-server-client-examples/main.js:84]: We should print the Service IP for the VM.

Comment (salanki) [reply virtual-server/examples/nodejs/virtual-server-client-examples/main.js:84]: The Service IP of the VM, as well as Internal IP should be set as statuses on the VirtualServer CRD so one does not have to look in a bunch of other objects to get them.

**Comment by goock** [comment]
From my perspective, it could be better to change paths for the examples from:
```
virtual-server/examples/<lang>/virtual-server-client-examples/
```
to

```
virtual-server/examples/virtual-server-client/<lang>
```

File: virtual-server/examples/nodejs/virtual-server-client-examples/client.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+      return new Promise((resolve, reject) => {
+        stream.on('data', async e => {
+          if(e.type === 'DELETED') {
+            stream.end()
+            return reject(new Error(`VirtualServer ${namespace}/${name} was deleted`))
+          }
+          else {
+            if(!!e.object.status) {
+              const readyCondition = e.object.status.conditions.filter(c => c.type === 'Ready')[0] || {}
+              if(readyCondition.status === 'True') {
```
Comment (goock) [inline virtual-server/examples/nodejs/virtual-server-client-examples/client.js:124]: Despite proper vs status:
```
NAME                STATUS               REASON               STARTED
my-virtual-server   VirtualServerReady   VirtualServerReady   True
```
the condition does not give us working VMI:
```
  Conditions:
    Last Transition Time:  2021-03-05T10:16:02Z
    Message:               VirtualServerReady
    Reason:                VirtualServerReady
    Status:                True
    Type:                  Ready
    Last Transition Time:  2021-03-05T10:16:02Z
    Message:               ServicesReady
    Reason:                ServicesReady
    Status:                True
    Type:                  ServicesReady
    Last Transition Time:  2021-03-05T10:16:02Z
    Message:               0/1384 nodes are available: 1162 Insufficient nvidia.com/gpu ...
    Reason:                Pending
    Status:                False
    Type:                  VirtualMachineReady
    Last Transition Time:  2021-03-05T10:16:20Z
    Message:               VirtualServerStarted
    Reason:                VirtualServerStarted
    Status:                True
    Type:                  VirtualServerStarted
  Network:
    Tcp IP:  207.53.234.148

```
Is this OK?

Comment (YitzyD) [reply virtual-server/examples/nodejs/virtual-server-client-examples/client.js:124]: > 0/1384 nodes are available: 1162 Insufficient nvidia.com/gpu ...
>     Reason:                Pending
>     Status:                False
>     Type:                  VirtualMachineReady

This has been fixed in recent commit

File: virtual-server/examples/nodejs/virtual-server-client-examples/main.js (hunk 1 above)
Comment (YitzyD) [reply virtual-server/examples/nodejs/virtual-server-client-examples/main.js]: done: https://github.com/coreweave/kubernetes-cloud/pull/7/commits/d26b26251ed1d193896dd649d4fa7e35562c4da0

File: virtual-server/examples/nodejs/virtual-server-client-examples/main.js (hunk 2 above)
Comment (YitzyD) [reply virtual-server/examples/nodejs/virtual-server-client-examples/main.js]: done: https://github.com/coreweave/kubernetes-cloud/pull/7/commits/d26b26251ed1d193896dd649d4fa7e35562c4da0

**Comment by YitzyD** [comment]
> From my perspective, it could be better to change paths for the examples from:
> 
> ```
> virtual-server/examples/<lang>/virtual-server-client-examples/
> ```
> 
> to
> 
> ```
> virtual-server/examples/virtual-server-client/<lang>
> ```
Agreed.
done: https://github.com/coreweave/kubernetes-cloud/pull/7/commits/4067e2e9e0587d2897089a49c131812cd9fdb6a9

**Comment by salanki** [comment]
Can one of you please add a small generic README.md with what it does, and how to run, where it fetches the kubeconfig from and that `client.js` can be included in their own application as a library.

**Comment by YitzyD** [comment]
> Can one of you please add a small generic README.md with what it does, and how to run, where it fetches the kubeconfig from and that `client.js` can be included in their own application as a library.

https://github.com/coreweave/kubernetes-cloud/blob/yd/vs-nodejs-example/virtual-server/examples/nodejs/Readme.md

**Comment by YitzyD** [comment]
Added a README and updated to meet current VS spec

File: virtual-server/examples/nodejs/Readme.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
+
+## Virtual Server NodeJS Example
+
+An example implementation in NodeJS of a kubernetes client to interact with a Virtual Server resource on Coreweave Cloud.
```
Comment (brianventuro) [inline virtual-server/examples/nodejs/Readme.md:4]: CoreWeave

File: virtual-server/examples/nodejs/Readme.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+The example is broken down into two parts, the [client](client.js), which is glue code layered on top of the godaddy implementation of a [kubernetes-client](https://github.com/godaddy/kubernetes-client), and the [application](main.js); both of which are described below.
+
+### Client
+The client ([client.js](client.js)) provides an interface to create a kubernetes client using your kubernetes credentials (token). Client.js may be dropped into any application and used to interface with Virtual Servers. A set of functions specific to creating, modifying, and checking the status of a Virtual Server are additionally provided and operate as follows:
+> All functions below are within the client.virtualServer object
+
+`init()`: Initialize the client
+> *init must be called prior to using the Client.*
+> 
+> *init will set up the client so that it is aware of Coreweave and Kubevirt CRDs as well as the kubevirt subresource api.*
```
Comment (brianventuro) [inline virtual-server/examples/nodejs/Readme.md:17]: CoreWeave

File: virtual-server/examples/nodejs/Readme.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+`get({namespace, name})`: Retrieves a Virtual Server
+* `namespace`: Namespace wherein the Virtual Server is deployed
+* `name`: Name of the Virtual Server
+
+`list({namespace})`: List all Virtual Servers
+* `namespace`: Namespace wherein the Virtual Servers is deployed
```
Comment (brianventuro) [inline virtual-server/examples/nodejs/Readme.md:24]: singular Virtual Server

File: virtual-server/examples/nodejs/Readme.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+`create(manifest)`: Retrieves a Virtual Server
```
Comment (brianventuro) [inline virtual-server/examples/nodejs/Readme.md:26]: Creates a Virtual Server?

File: virtual-server/examples/nodejs/Readme.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+`start({namespace, name})`: Start a stopped Virtual Server
+* `namespace`: Namespace wherein the Virtual Server is deployed
+* `name`: Name of the Virtual Server
+
+`stop({namespace, name})`: Stop a running Virtual Server
+* `namespace`: Namespace wherein the Virtual Server is deployed
+* `name`: Name of the Virtual Server
+
+### Application
+The application ([main.js](main.js)) implements client.js in order to create, watch, get, and update an example Virtual Server. The application serves as an example of how to implement client.js. Additionally, the application illustrates how one might retrieve status information from a running Virtual Server, retrieving it's run state and network information. Finally, the example shows how to simply interface with the subresource api, starting and stopping a Virtual Server.
```
Comment (brianventuro) [inline virtual-server/examples/nodejs/Readme.md:51]: "its" is the possessive form of it, not "it's"

File: virtual-server/examples/nodejs/Readme.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+  },
+  spec: {
+    region: "ORD1",
+    os: {
+      type: "linux"
+    },
+    resources: {
+      gpu: {
+        type: "Quadro_RTX_4000",
+        count: 4
```
Comment (brianventuro) [inline virtual-server/examples/nodejs/Readme.md:70]: Let's change this example to 1 GPU.

File: virtual-server/examples/nodejs/main.js [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+// Configure a sample spec
+virtualServerManifest.spec = {
+  region: "ORD1",
+  os: {
+    type: "linux"
+  },
+  resources: {
+    gpu: {
+      type: "Quadro_RTX_4000",
+      count: 4
```
Comment (brianventuro) [inline virtual-server/examples/nodejs/main.js:18]: Please set to 1 for example.

File: virtual-server/examples/nodejs/main.js [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    },
+    cpu: {
+      count: 2
```
Comment (brianventuro) [inline virtual-server/examples/nodejs/main.js:21]: This should really be set to our standard config:

3 cpu
16 gb ram

File: virtual-server/examples/nodejs/Readme.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+      },
+      cpu: {
+        count: 2
```
Comment (brianventuro) [inline virtual-server/examples/nodejs/Readme.md:73]: set to 3 cpu to match standard config

File: virtual-server/examples/nodejs/main.js [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+  /* 
+  Sample Output:
+    InternalIP: 1.2.3.4
+    ExternalIP: 0.0.0.0
+    FloatingIPs:
+        sample-floating-ip-service: 1.1.1.1
+  */
+
+  // Increase the memory resource request to 32Gi and update the VirtualServer
+  vs.spec.resources.memory = "32Gi"
```
Comment (brianventuro) [inline virtual-server/examples/nodejs/main.js:126]: Should this just be an example snippet outside of this example?

Work under the idea that every single person who uses this example is just going to run this to start.

File: virtual-server/examples/nodejs/util.js [hunk 14]
```diff
@@ -0,0 +1,67 @@
+// Validates is a quantity is a valid k8s resource.Quantity
```
Comment (brianventuro) [inline virtual-server/examples/nodejs/util.js:1]: Not english

**Review by brianventuro — CHANGES_REQUESTED** [review]

**Pull request merged**